proof/invariant-abstract/DetSchedInvs_AI.thy

```diff
@@ -3009,8 +3017,58 @@ lemma active_reply_scsE:
   shows "active_reply_scs_2 replies' refill_cfgs'"
   using assms by (simp add: active_reply_scs_2_def active_if_reply_sc_at_2_def)
 
+abbreviation priority_ordered :: "obj_ref list \<Rightarrow> (obj_ref \<rightharpoonup> priority) \<Rightarrow> bool" where
+  "priority_ordered ts prios \<equiv> sorted_wrt (img_ord prios (opt_ord_R (\<ge>))) ts"
+
+definition sorted_ipc_queues_except_set_2 ::
+  "obj_ref set \<Rightarrow> (obj_ref \<rightharpoonup> obj_ref list) \<Rightarrow> (obj_ref \<rightharpoonup> obj_ref list) \<Rightarrow> (obj_ref \<rightharpoonup> priority) \<Rightarrow> bool" where
+  "sorted_ipc_queues_except_set_2 except ep_qs ntfn_qs prios \<equiv>
+   \<forall>ptr. ptr \<notin> except \<longrightarrow>
+           (none_top (\<lambda>q. priority_ordered q prios) (ep_qs ptr)
+            \<and> none_top (\<lambda>q. priority_ordered q prios) (ntfn_qs ptr))"
+
+abbreviation sorted_ipc_queues_except_set :: "obj_ref set \<Rightarrow> 'z::state_ext state \<Rightarrow> bool" where
+  "sorted_ipc_queues_except_set except s \<equiv>
+     sorted_ipc_queues_except_set_2 except (ep_queues_of s) (ntfn_queues_of s) (prios_of s)"
+
+abbreviation "sorted_ipc_queues_except_2 t \<equiv> sorted_ipc_queues_except_set_2 {t}"
+abbreviation "sorted_ipc_queues_except t s \<equiv> sorted_ipc_queues_except_2 t (ep_queues_of s) (ntfn_queues_of s) (prios_of s)"
+abbreviation "sorted_ipc_queues_2 \<equiv> sorted_ipc_queues_except_set_2 {}"
+abbreviation "sorted_ipc_queues s \<equiv> sorted_ipc_queues_2 (ep_queues_of s) (ntfn_queues_of s) (prios_of s)"
+
+lemmas sorted_ipc_queues_except_set_def = sorted_ipc_queues_except_set_2_def
+lemmas sorted_ipc_queues_except_def = sorted_ipc_queues_except_set_2_def
+lemmas sorted_ipc_queues_def = sorted_ipc_queues_except_set_2_def
+
+lemma sorted_ipc_queues_endpoint_priority_ordered:
+  "\<lbrakk>sorted_ipc_queues s; ep_queues_of s ptr = Some q\<rbrakk> \<Longrightarrow> priority_ordered q (prios_of s)"
+  by (fastforce simp: sorted_ipc_queues_def dest: spec[where x=ptr])
+
+lemma sorted_ipc_queues_notification_priority_ordered:
+  "\<lbrakk>sorted_ipc_queues s; ntfn_queues_of s ptr = Some q\<rbrakk> \<Longrightarrow> priority_ordered q (prios_of s)"
+  by (fastforce simp: sorted_ipc_queues_def dest: spec[where x=ptr])
+
+declare opt_map_Some_comp[simp del]
```

Comment:
I did decide in AInvs to remove `opt_map_Some_comp` from the simp set, but it ended up back in the simp set during Refine (in a way I haven't tracked down yet). I think it would be best to not have it in the simp set at all.

Reply:
That looks painful. I'm not really an expert on the construction and deployment of these, so I'll leave it to @lsf37. I can say that it should be possible to track down where it gets reintroduced in Refine and kick it out early. Maybe in Invariants_H?